crates/miden-objects/src/transaction/tx_args.rs

```diff
@@ -243,15 +250,53 @@ impl TransactionScript {
 
     /// Returns a new [TransactionScript] instantiated from the provided components.
     ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The provided `script_args_key` is not presented in the `inputs` map.
+    ///
     /// # Panics
     /// Panics if the specified entrypoint is not in the provided MAST forest.
     pub fn from_parts(
         mast: Arc<MastForest>,
         entrypoint: MastNodeId,
         inputs: BTreeMap<Digest, Vec<Felt>>,
-    ) -> Self {
+        script_args_key: Option<Digest>,
+    ) -> Result<Self, TransactionScriptError> {
         assert!(mast.get_node_by_id(entrypoint).is_some());
-        Self { mast, entrypoint, inputs }
+
+        // check that provided `script_args_key` is presented in the `inputs` map
+        if let Some(args_key) = script_args_key {
+            if !inputs.contains_key(&args_key) {
+                return Err(TransactionScriptError::InvalidScriptArgsKey(args_key));
+            }
+        }
+
+        Ok(Self {
+            mast,
+            entrypoint,
+            inputs,
+            script_args_key,
+        })
+    }
+
+    // MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Sets the `script_args_key` to the commitment of the provided arguments array and extends
+    /// the `inputs` map with the `COMPUTED_COMMITMENT -> [[script_args]]` entry.
+    ///
+    /// Script arguments is an optional array of [`Felt`]s which could be easily accessed at the
+    /// beginning of the transaction script execution. Commitment of this array (`script_args_key`)
+    /// is automatically pushed to the operand stack at the beginning of the transaction script
+    /// execution, so user can get access the underlying arguments array just using the
+    /// `adv.push_mapval` and `adv_push.n` instructions.
+    ///
+    /// See `test_tx_script_args` test as an example of their usage.
+    pub fn with_script_args(mut self, script_args: &[Felt]) -> Self {
+        let args_key = Hasher::hash_elements(script_args);
+        self.inputs.insert(args_key, script_args.to_vec());
+        self.script_args_key = Some(args_key);
+        self
```

Comment:
It may be worth checking here that we re not overriding some existing value in the `inputs` map (i.e., if the map has a different value for the same key in it already, we should probably return an error).

A bigger question is whether we should just set the key here. This is how it works for notes, and we are being somewhat inconsistent about how note and tx script args work. But probably it is fine to leave as is for now.